### ISSUE TITLE
Could not connect to Azure SQL Database on version 17.8.1.1

### ISSUE THREAD

**Issue by bpolgar-waracle** [microsoft/homebrew-mssql-release#59]
## Description

We have an Azure SQL Database running and a Python script connecting to it using `pyodbc` / `msodbcsql17` / `mssql-tools`

Latter two have been installed via brew, without any version specified, and run on a [Microsoft hosted agent](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md) `macOS-10.15`

The following approach was used to install the related driver on the agent, detailed by this [doc page](https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver15):

    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
    brew tap microsoft/mssql-release https://github.com/Microsoft/homebrew-mssql-release
    brew update
    HOMEBREW_NO_ENV_FILTERING=1 ACCEPT_EULA=Y brew install msodbcsql17 mssql-tools

We have been getting the following error when trying to connect to the database with this connection string:

- Connection string:
  - `DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<db-host>;DATABASE=<db-name>;UID=<db-user-name>;PWD=<db-user-password>;`

- Error message:
  - `OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection (0) (SQLDriverConnect)')`

## Attempted fixes

We have tried a few solutions, but without success:

- `Minimum TLS Version` on all of 1.0, 1.1, 1.2
- Installing previous versions of `msodbcsql17` / `mssql-tools`

## Successful workaround

- Symlinking `/usr/local/Cellar/openssl@1.1/1.1.1l` to `openssl` instead of the current 3.0 version

**Comment by bpolgar-waracle** [comment]
Here is a trace from a failed connection:

    [ODBC][12271][1632736389.150141][__handles.c][460]
		    Exit:[SQL_SUCCESS]
			    Environment = 0x7f847e01fe00
    [ODBC][12271][1632736389.150280][SQLSetEnvAttr.c][189]
		    Entry:
			    Environment = 0x7f847e01fe00
			    Attribute = SQL_ATTR_ODBC_VERSION
			    Value = 0x3
			    StrLen = 4
    [ODBC][12271][1632736389.150397][SQLSetEnvAttr.c][381]
		    Exit:[SQL_SUCCESS]
    [ODBC][12271][1632736389.150526][SQLAllocHandle.c][377]
		    Entry:
			    Handle Type = 2
			    Input Handle = 0x7f847e01fe00
		    UNICODE Using encoding ASCII 'UTF-8' and UNICODE 'UCS-2-INTERNAL'
    
    [ODBC][12271][1632736389.154954][SQLAllocHandle.c][513]
		    Exit:[SQL_SUCCESS]
			    Output Handle = 0x7f847e80d600
    [ODBC][12271][1632736389.155817][SQLDriverConnectW.c][290]
		    Entry:
			    Connection = 0x7f847e80d600
			    Window Hdl = 0x0
			    Str In = [DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<db-host>;DATABASE=<db-name>;UID=g...][length = 171 (SQL_NTS)]
			    Str Out = 0x0
			    Str Out Max = 0
			    Str Out Ptr = 0x0
			    Completion = 0
    [ODBC][12271][1632736389.356331][__handles.c][460]
		    Exit:[SQL_SUCCESS]
			    Environment = 0x7f847e812800
    [ODBC][12271][1632736389.356526][SQLGetEnvAttr.c][157]
		    Entry:
			    Environment = 0x7f847e812800
			    Attribute = 65002
			    Value = 0x7ffeee7b5b70
			    Buffer Len = 128
			    StrLen = 0x7ffeee7b5b54
    [ODBC][12271][1632736389.356659][SQLGetEnvAttr.c][273]
		    Exit:[SQL_SUCCESS]
    [ODBC][12271][1632736389.356799][SQLFreeHandle.c][220]
		    Entry:
			    Handle Type = 1
			    Input Handle = 0x7f847e812800
    [ODBC][12271][1632736389.357062][SQLDriverConnectW.c][699]
		    Exit:[SQL_ERROR]
    [ODBC][12271][1632736389.357236][SQLDriverConnect.c][748]
		    Entry:
			    Connection = 0x7f847e80d600
			    Window Hdl = 0x0
			    Str In = [DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<db-host>;DATABASE=<db-name>;UID=gc...][length = 171 (SQL_NTS)]
			    Str Out = 0x7ffeee7b6530
			    Str Out Max = 2048
			    Str Out Ptr = 0x0
			    Completion = 0
		    DIAG [08001] [Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection
    
		    DIAG [01S00] [Microsoft][ODBC Driver 17 for SQL Server]Invalid connection string attribute
    
    [ODBC][12271][1632736389.458030][SQLDriverConnect.c][1637]
		    Exit:[SQL_ERROR]
    [ODBC][12271][1632736389.458256][SQLGetDiagRecW.c][535]
		    Entry:
			    Connection = 0x7f847e80d600
			    Rec Number = 1
			    SQLState = 0x7ffeee7b88d4
			    Native = 0x7ffeee7b80c4
			    Message Text = 0x7ffeee7b80d0
			    Buffer Length = 1023
			    Text Len Ptr = 0x7ffeee7b80ce
    [ODBC][12271][1632736389.462939][SQLGetDiagRecW.c][596]
		    Exit:[SQL_SUCCESS]
			    SQLState = [08001]
			    Native = 0x7ffeee7b80c4 -> 0 (32 bits)
			    Message Text = [[Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection]
    [ODBC][12271][1632736389.463220][SQLFreeHandle.c][290]
		    Entry:
			    Handle Type = 2
			    Input Handle = 0x7f847e80d600
    [ODBC][12271][1632736389.463483][SQLFreeHandle.c][339]
		    Exit:[SQL_SUCCESS]

**Comment by v-chojas** [comment]
Could you try a connection with `isql -v -k '<connection string goes here>'` to show the full set of error messages? It sounds like you found the problem was due to OpenSSL, so could you also show output of `ls -l /usr/local/opt/openssl/lib` ?

**Comment by jh88** [comment]
I had the same error. 
When you `brew install msodbcsql17 mssql-tools`, it also installs `openssl` and by default it refers to `openssl@3`, however you need `openssl@1.1`.
You can install 1.1 and change the symbolic link
```
brew install openssl@1.1

# you might need to delete the old symlink first
# rm /usr/local/opt/openssl

ln -s /usr/local/Cellar/openssl@1.1/1.1.1l /usr/local/opt/openssl
```
Currently the latest version openssl 1.1 is `1.1.1l`, for future reference you might need to change directory accordingly.

**Comment by bpolgar-waracle** [comment]
@v-chojas 

> Could you try a connection with `isql -v -k '<connection string goes here>'` to show the full set of error messages? It sounds like you found the problem was due to OpenSSL, so could you also show output of `ls -l /usr/local/opt/openssl/lib` ?

    ➜  ~ isql -v -k "DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<db-host>;DATABASE=<db-name>;UID=<db-user-name>;PWD=<db-user-password>;"  
    [08001][Microsoft][ODBC Driver 17 for SQL Server]SSL Provider: [OpenSSL library could not be loaded, make sure OpenSSL 1.0 or 1.1 is installed] 
    [08001][Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection 
    [ISQL]ERROR: Could not SQLDriverConnect 

    ➜  ~ ls -l /usr/local/opt/openssl/lib 
    total 22544 
    drwxr-xr-x  5 user-name  admin      160  7 Sep 12:46 engines-3 
    -rw-r--r--  1 user-name  admin  3627768 25 Sep 00:20 libcrypto.3.dylib 
    -r--r--r--  1 user-name  admin  6498816  7 Sep 12:46 libcrypto.a 
    lrwxr-xr-x  1 user-name  admin       17  7 Sep 12:46 libcrypto.dylib -> libcrypto.3.dylib 
    -r--r--r--  1 user-name  admin   578568 25 Sep 00:20 libssl.3.dylib 
    -r--r--r--  1 user-name  admin   828120  7 Sep 12:46 libssl.a 
    lrwxr-xr-x  1 user-name  admin       14  7 Sep 12:46 libssl.dylib -> libssl.3.dylib 
    drwxr-xr-x  3 user-name  admin       96  7 Sep 12:46 ossl-modules 
    drwxr-xr-x  5 user-name  admin      160 25 Sep 00:20 pkgconfig 

@jh88 I did manage to get the connection working with using `1.1.1l` version of openssl

**Comment by bpolgar-waracle** [comment]
So the `1.1` openssl is not a workaround, but the expected version to use?

**Comment by v-chojas** [comment]
Yes. As the message says, "make sure OpenSSL 1.0 or 1.1 is installed". The driver was released before OpenSSL 3.0(!) even became available. It looks like the unversioned links still point to the libs despite them probably not being compatible.

**Comment by tessa-hudson** [comment]
Can confirm that this also worked for me! I was getting the same error message when using pyodbc with ODBC Driver 17 for SQL Server:

OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection (0) (SQLDriverConnect)')

**Comment by tswayne** [comment]
Hey, yeah also went through all of this - something is screwy with the cask deps that's causing this.  When I type `brew deps -n msodbcsql17` I see `openssl@3`, where some of my teammates (and I used to) see `openssl@1.1`.  So now whenever I install `msodbcsql17` out of the box, it overwrites openssl.  

Is that dependency a product of this repo, or is there a different place that is configured/needs to be fixed?

**Comment by v-chojas** [comment]
You can see here, we do not depend on a specific version of openssl: bhttps://github.com/microsoft/homebrew-mssql-release/blob/master/Formula/msodbcsql17.rb
But if 1.0 or 1.1 is installed, the driver will use it.  We are looking into what is different with this new version of openssl.

**Comment by TheCaffinatedDeveloper** [comment]
Followed the steps to install openssl@1.1 and removed/created the symlink to 1.1 but still seeing the same error.
```
isql -v -k '<connection string goes here>'

[08001][Microsoft][ODBC Driver 17 for SQL Server]SSL Provider: [OpenSSL library could not be loaded, make sure OpenSSL 1.0 or 1.1 is installed]
[08001][Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection
[ISQL]ERROR: Could not SQLDriverConnect
```

```
which openssl
/usr/local/Cellar/openssl@1.1/1.1.1l_1/bin/openssl
```

```
ls -la /usr/local/opt
...
36 Oct 12 11:04 openssl -> /usr/local/Cellar/openssl@1.1/1.1.1l
...
```

**Comment by hteo1337** [comment]
hello,
this is how I solved the issue (following this comment https://github.com/mkleehammer/pyodbc/issues/965#issuecomment-937778399):
1. I removed the link in /usr/local/opt/:
`rm openssl`
2. Created a new link to openssl@1.1
`ln -s /usr/local/Cellar/openssl@1.1/1.1.1l_1 /usr/local/opt/openssl`
I'm using pycharm + flask+pyodbc and Azure SQL DB.
This was the error received: 
OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection (0) (SQLDriverConnect)')

**Comment by achokey-crp** [comment]
> Ah I managed to solve it thanks to you @v-chojas. I only had `openssl@1.1` in `/opt/homebrew/opt`
> 
> I created a symlink this way:
> 
> `ln -s /opt/homebrew/opt/openssl@1.1 /opt/homebrew/opt/openssl`
> 
> and now everything works as expected !

This worked for me as well. Thanks for this!

**Comment by FatmaKursun** [comment]
> ## Description
> We have an Azure SQL Database running and a Python script connecting to it using `pyodbc` / `msodbcsql17` / `mssql-tools`
> 
> Latter two have been installed via brew, without any version specified, and run on a [Microsoft hosted agent](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md) `macOS-10.15`
> 
> The following approach was used to install the related driver on the agent, detailed by this [doc page](https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver15):
> 
> ```
> /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
> brew tap microsoft/mssql-release https://github.com/Microsoft/homebrew-mssql-release
> brew update
> HOMEBREW_NO_ENV_FILTERING=1 ACCEPT_EULA=Y brew install msodbcsql17 mssql-tools
> ```
> 
> We have been getting the following error when trying to connect to the database with this connection string:
> 
> * Connection string:
>   
>   * `DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<db-host>;DATABASE=<db-name>;UID=<db-user-name>;PWD=<db-user-password>;`
> * Error message:
>   
>   * `OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Client unable to establish connection (0) (SQLDriverConnect)')`
> 
> ## Attempted fixes
> We have tried a few solutions, but without success:
> 
> * `Minimum TLS Version` on all of 1.0, 1.1, 1.2
> * Installing previous versions of `msodbcsql17` / `mssql-tools`
> 
> ## Successful workaround
> * Symlinking `/usr/local/Cellar/openssl@1.1/1.1.1l` to `openssl` instead of the current 3.0 version

Hello, I worked on this issue for a very long time and this solution worked for me.


export DYLD_LIBRARY_PATH =/opt/homebrew/Cellar/openssl@1.1/1.1.1l_1/lib/: $DYLD_LIBRARY_PATH
echo $DYLD_LIBRARY_PATH

you need to add your own directory that openssl1.1

**Comment by cbrunnkvist** [comment]
Yes @FatmaKursun you got one space too much after the colon separator in your paste but otherwise good suggestion, that worked for me 🤩 

    export DYLD_LIBRARY_PATH =/opt/homebrew/Cellar/openssl@1.1/1.1.1l_1/lib/:$DYLD_LIBRARY_PATH
    echo $DYLD_LIBRARY_PATH

**Comment by cbrunnkvist** [comment]
Tangential but perhaps a timesaver for anyone attempting the symlink workaround: on macOS/BSD at least, note that `ln` might cause you to inadvertently shoot yourself in the foot: **if** `/usr/local/opt/openssl` is already a symlink, the command in the as quoted will first **follow** the existing symlink first and create a new one named `1.1.1l_1` __inside__ of the target directory (e.g. in in the 3.3 one) instead of overwriting it with a symlink itself... 

You'll need to `ln -shf /usr/local/Cellar/openssl@1.1/1.1.1l_1 /usr/local/opt/openssl` to get the correct result.

From `man ln`:
```man
     ln [-Ffhinsv] source_file [link_name]

     -h    If the link_name or link_dirname is a symbolic link, do not follow it.  This is most useful with
           the -f option, to replace a symlink which may point to a directory.
```

**Comment by mmoye** [comment]
@cbrunnkvist I tried 8 bazillion things and your suggestion was the only thing that finally worked. THANK YOU.

**Comment by palkim** [comment]
> Tangential but perhaps a timesaver for anyone attempting the symlink workaround: on macOS/BSD at least, note that `ln` might cause you to inadvertently shoot yourself in the foot: **if** `/usr/local/opt/openssl` is already a symlink, the command in as quoted will actually **follow** the existing symlink first and create a new one named `1.1.1l_1` **inside** of the target directory (e.g. in in the 3.3 one) instead of overwriting it with a symlink itself...
> 
> You'll need to `ln -shf /usr/local/Cellar/openssl@1.1/1.1.1l_1 /usr/local/opt/openssl` to get the correct result.
> 
> From `man ln`:
> 
> ```
>      ln [-Ffhinsv] source_file [link_name]
> 
>      -h    If the link_name or link_dirname is a symbolic link, do not follow it.  This is most useful with
>            the -f option, to replace a symlink which may point to a directory.
> ```

many many thanks, it worked!

**Comment by palkim** [comment]
cbrunnkvist many thanks!

**Comment by agouliel** [comment]
On my M1 iMac running Monterey, I had the following situation in /opt/homebrew/opt:

openssl@1.1 -> ../Cellar/openssl@1.1/1.1.1l_1
openssl@3 -> ../Cellar/openssl@3/3.0.0_1
openssl -> ../Cellar/openssl@3/3.0.0_1

So the plain "openssl" link was directing to version 3. After seeing the answer of @PILLOWPET, I did the following:

mv openssl opensslbak
ln -s ../Cellar/openssl@1.1/1.1.1l_1 openssl

So now the situation is this:

openssl -> ../Cellar/openssl@1.1/1.1.1l_1
openssl@1.1 -> ../Cellar/openssl@1.1/1.1.1l_1
openssl@3 -> ../Cellar/openssl@3/3.0.0_1
opensslbak -> ../Cellar/openssl@3/3.0.0_1

And now it works.

**Comment by Niladri-B** [comment]
I ended up in the same problem after upgrading from macOS Catalina > macOS Monterey with Python 3.8.9

To resolve the issue at my end, here are the steps I needed to take:

1. uninstall brew https://github.com/homebrew/install#uninstall-homebrew     
2. Install brew + msodbcsql17 as detailed here: https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver15. 
3. Remove v3 of openssl that gets installed in the aforementioned step as mentioned by @dianabursac2019 here https://github.com/microsoft/homebrew-mssql-release/issues/59#issuecomment-953111934
`brew uninstall --force openssl@3 `  ` brew uninstall --ignore-dependencies openssl`
4.  Remove other files and folders related to v3 manually `rm -rf /usr/local/etc/openssl@3  `
5.  Perform a brew cleanup `brew cleanup -s openssl@3  `
6.  Install v1 of openssl `brew install openssl@1.1`
7.  Run brew link on the above `brew link openssl@1.1 `, there might be some warnings/errors but it gets resolved following step 8
8. Add to $PATH variable  _all_ of the following for the `zsh`
` echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc `
`echo 'export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"' >> ~/.zshrc  `
`echo 'export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"' >> ~/.zshrc      `
9. Restart Terminal and confirm $PATH contains the above paths by running `echo $PATH`
10.  Create symlink as suggested by @cbrunnkvist https://github.com/microsoft/homebrew-mssql-release/issues/59#issuecomment-963934545  `ln -shf /usr/local/Cellar/openssl@1.1/1.1.1l_1 /usr/local/opt/openssl`
11.  Add the path to DYLD_LIBRARY_PATH as suggested by @FatmaKursun  and @cbrunnkvist above. However I needed to add it directly to my `.zshrc`file as `echo 'export DYLD_LIBRARY_PATH=/opt/homebrew/Cellar/openssl@1.1/1.1.1l_1/lib/:$DYLD_LIBRARY_PATH' >> ~/.zshrc`
12.  Restart Terminal and confirm above was successfully added as  `echo $DYLD_LIBRARY_PATH`
13.  Run the `iclv`command and now it should work to connect to Azure SQL Database.     
`DYLD_PRINT_LIBRARIES=1 isql -v -k "DRIVER={ODBC Driver 17 for SQL Server};SERVER=tcp:<server_address>;DATABASE=<database_name>;UID=<user>;PWD=<123>;"`

**Comment by rpus** [comment]
@v-chojas you say:

> You can see here, we do not depend on a specific version of openssl: bhttps://github.com/microsoft/homebrew-mssql-release/blob/master/Formula/msodbcsql17.rb
> But if 1.0 or 1.1 is installed, the driver will use it. We are looking into what is different with this new version of openssl.

... but I don't think that's quite right; you can see [here](https://formulae.brew.sh/formula/openssl@3) that `openssl` is now an *alias* for `openssl@3` (since v3 is now the latest version, `openssl@3` is "also known as" `openssl` in the context of Homebrew), so it seems your two packages are inadvertently nowadays installing "dependencies" which can then break them at runtime!

I think the simplest thing (guaranteed to work safely?) is to change the installation scripts for [`msodbcsql`](https://github.com/microsoft/homebrew-mssql-release/blob/master/Formula/msodbcsql.rb#L11) and [`mssql-tools`](https://github.com/microsoft/homebrew-mssql-release/blob/master/Formula/mssql-tools.rb#L11) to depend on [`openssl@1.1`](https://formulae.brew.sh/formula/openssl@1.1).

**Comment by v-chojas** [comment]
We are not going to change history, but the next release we make will add support for OpenSSL 3.0.

**Comment by shaneburrell** [comment]
Has anyone just tried a local edit to home-brew formulas to specify openssl@1.1?

**Comment by shaneburrell** [comment]
Ok that seems to fix the dependency.  But now.  
- php/8.1.0/bin/php is loading libcrypto in an unsafe way
and
- WARNING: /opt/homebrew/bin/isql is loading libcrypto in an unsafe way

**Comment by rpus** [comment]
@shaneburrell I actually just tried that too, and I also get `WARNING: /opt/homebrew/bin/sqlcmd is loading libcrypto in an unsafe way` on startup (followed by an immediate `Canceling connection...` and exit, so perhaps changing the dependency is not a fix anyway).

**Comment by shaneburrell** [comment]
I think it may be as simple as creating a openssl link to openssl@1.1 .  I'll try it in a few but it may just be that since using a versioned lib dir it may need a ln to load in standard directory.

**Comment by shaneburrell** [comment]
All suggestions at this point seem to lead to:
WARNING: /opt/homebrew/bin/isql is loading libcrypto in an unsafe way

**Comment by spencerwilson** [comment]
> we are not going to change history

@v-chojas This is commendable! Although I’m not sure @rpus was suggesting changing already-published formula versions.

Rather, publish a new formula version that just pins the openssl dependency to the versioned formula. Would that “stop the bleeding” in this issue and buy the maintainers more time to work updating the software to be compatible with v3?

**Comment by shaneburrell** [comment]
Ok so the cleanest solution is to correctly specify the right openssl@1.1 in the brew formulas. This resolves the packages download @3.  The packages are hitting lib crypto from apple and not home-brew. I'm not sure how the actual binaries determine which dylib to use but exporting DYLD_LIBRARY_PATH with /opt/homebrew/Lib/ fixes the issue and I can now connect on M1 based Monterey with just three steps.

- Set formulas to openssl@1.1 manually
- export DYLD_LIBRARY_PATH
- install mssql brew formula normally

Would it be possible to bump a version and specify the intended working openssl@1.1 in the formulas moving forward until openssl@3 is actually supported?

**Comment by shaneburrell** [comment]
> @shaneburrell I actually just tried that too, and I also get `WARNING: /opt/homebrew/bin/sqlcmd is loading libcrypto in an unsafe way` on startup (followed by an immediate `Canceling connection...` and exit, so perhaps changing the dependency is not a fix anyway).

@rpus If you add the DYLD_LIBRARY_PATH so it can find the correct libcrypto it solves the issue.

**Comment by v-chojas** [comment]
The next release will support OpenSSL 3.0. As many of you have now discovered, even if you have 1.1 installed you still need the additional steps described earlier in this issue to have it work.

**Comment by a087861** [comment]
Any idea on the ETA of the next release?

**Comment by v-chojas** [comment]
Early 2022.

**Comment by alfredoyong** [comment]
> I had the same error. When you `brew install msodbcsql17 mssql-tools`, it also installs `openssl` and by default it refers to `openssl@3`, however you need `openssl@1.1`. You can install 1.1 and change the symbolic link
> 
> ```
> brew install openssl@1.1
> 
> # you might need to delete the old symlink first
> # rm /usr/local/opt/openssl
> 
> ln -s /usr/local/Cellar/openssl@1.1/1.1.1l /usr/local/opt/openssl
> ```
> 
> For future reference, you might need to change directory accordingly. Currently the latest version of openssl 1.1 is `1.1.1l`

Yes, this worked for me. Today, current version 1 is openSSL@1.1/1.1.1m, so that's the one I linked:
```
 ln -s /usr/local/Cellar/openssl@1.1/1.1.1m /usr/local/opt/openssl
 ```

This article explains exactly how to properly install openSSL with Brew, what is a "keg-only" and why you need to manually link (because the mac comes with freeSSL)

https://yasar-yy.medium.com/installing-openssl-library-on-macos-catalina-6777a2e238a6

**Comment by wivaku** [comment]
Instead of linking to the specific version (e.g. 1.1.1.m), you could link to the 1.1 link which links to the latest version.
```
rm -f /usr/local/opt/openssl && ln -s /usr/local/opt/openssl@1.1 /usr/local/opt/openssl
```

```
ls -ld /usr/local/opt/openssl*
/usr/local/opt/openssl ⇒ /usr/local/opt/openssl@1.1
/usr/local/opt/openssl@1.1 ⇒ ../Cellar/openssl@1.1/1.1.1m
/usr/local/opt/openssl@3 ⇒ ../Cellar/openssl@3/3.0.1
```

**Comment by dragnblow** [comment]
@David-Engel - is this where you want me to comment? 

So here you go..

Which OpenSSL returns the anaconda one - /opt/anaconda3/bin/OpenSSL 

And

OpenSSL version returns - OpenSSL 1.1.1l 24 Aug 2021.

Pyodbc.drivers returns the correct driver - odbc driver 17 for sql server.

The error:
08001 operational error: client unable to establish a connection sqldriverconnect.

Thank you!

**Comment by v-chojas** [comment]
He asked for the full set of error messages, not only an approximate retyping of one of them.

**Comment by dragnblow** [comment]
@v-chojas and @David-Engel 

i  have the entire error-screenshots here - github.com/mkleehammer/pyodbc/issues/994

**Comment by fenton-nick** [comment]
> Instead of linking to the specific version (e.g. 1.1.1.m), you could link to the 1.1 link which links to the latest version.
> 
> ```
> rm -f /usr/local/opt/openssl && ln -s /usr/local/opt/openssl@1.1 /usr/local/opt/openssl
> ```
> 
> ```
> ls -ld /usr/local/opt/openssl*
> /usr/local/opt/openssl ⇒ /usr/local/opt/openssl@1.1
> /usr/local/opt/openssl@1.1 ⇒ ../Cellar/openssl@1.1/1.1.1m
> /usr/local/opt/openssl@3 ⇒ ../Cellar/openssl@3/3.0.1
> ```

This worked for me.

**Comment by v-chojas** [comment]
msodbcsql18 with support for OpenSSL 3 has been released.
If you need msodbcsql17, its next version (17.9) will be released in a few days.

**Comment by Cap-JaTu** [comment]
I don't think ODBC Driver version 18 fixes the original problem. For both 17 and 18 I still need to:
```bash
ln -s /opt/homebrew/opt/openssl@1.1/lib/libssl.dylib /usr/local/lib/
ln -s /opt/homebrew/opt/openssl@1.1/lib/libcrypto.dylib /usr/local/lib/
```

to avoid _loading libcrypto in an unsafe way_ -warning & crash.

**Comment by v-chojas** [comment]
Please try again after removing any links you added,, remove `openssl@1.1` and install `openssl` (which should be the 3.0 version).

**Comment by Cap-JaTu** [comment]
Removing OpenSSL 1.1 isn't feasible as it is a dependency to other installed packages.

**Comment by v-chojas** [comment]
I mean to remove the links you added as a workaround before.

**Comment by Cap-JaTu** [comment]
Removing the links results in the original non-functional state.

**Comment by v-kevinwong** [comment]
I was able to test ODBC Driver 18 on a clean install with both `openssl` and `openssl@1.1` and that worked.

Are those of you having the issue still using driver 17.8?

**Comment by rpus** [comment]
I am fine now, and have only been using v17.

I was on v17.8 with workarounds (as previously on this thread).

I waited for the new v17 (i.e. v.17.9) Homebrew formulae to be released, and now I am fine once more (with no workarounds).

I did this:

- made sure my `DYLD_LIBRARY_PATH` env customisation was reverted/absent
- reverted my `brew edit` customisations (to `mssql-tools` and `msodbcsql17`, having previously forced then to depend on `openssl@1.1`)
- ran `brew uninstall` on those two MS Homebrew packages, to force an import/install of openssl v3 next...
- ran `brew install` on the same(-named) MS packages, to install v17.9 and the new openssl (v3) dependency
- carried on as usual (with no errors, and no customisations)

In other words, I am now finally "back to normal", with no customisations/edits, and not using any different(ly-named) packages/dependencies to what I was using originally before this issue was even raised.

To emphasise, I haven't needed to upgrade to (nor even to know about) v18...

**Comment by mucholove** [comment]
Hi, I'm getting the following error—not sure why. I have both openssl 1 & 3 installed via homebrew and am connecting to SQL Server 2014. Oddly enough, I can connect through Node.js, but PHP presents problems.

**Comment by v-chojas** [comment]
The driver is likely to be using OpenSSL 3 if you have it installed, the server may not support TLS 1.2, and you may have a configuration that prevents using older TLS versions. SQL 2014 has an update (see kb3135244 for more information) that adds TLS 1.2 support, but if you cannot change the server, you'll need to change the OpenSSL configuration to enable TLS 1.1 or even older (depending on the maximum version supported by server.)

**Comment by utkalRP** [comment]
@v-chojas I don't if it's correct forum or not but my isue is similar to this.
Server: Windows server 2012 R2 hosting SQL server 2012/2014/2016 with all latest patches+SPs enabling all TLS versions via registry
Client: Client is C++ Linux application compiled with OpenSSL 3.0.8 and aking ODBC connection via MS SQL Server ODBC Driver 17/18 with unixODBC manager 2.3.11.

When the client tries to connect sql server via ODBC driver with 'Client Hello', server doesnot respond with 'Server Hello'. Even openssl s_client is not able to connect to server on port 1433.

If we use older version of OpenSSL 1.1.1 it connects without any issue.

After extensive searching finally found reference to SSL_OP_IGNORE_UNEXPECTED_EOF, which has been introduced in OpenSSL 3.x and may be MS ODBC driver needs to set this option in SSL context (they know better their code). Link: https://github.com/microsoft/msphpsql/issues/1453#issuecomment-1595655486

Please help with latest SQL server ODBC driver with OpenSSL 3.0.8 suport.

**Comment by v-chojas** [comment]
> Even openssl s_client is not able to connect to server on port 1433.

That's not how it works - unless you have TDS8 enabled (which was only introduced in SQL 2022), the connection doesn't start with a TLS handshake.

What error message are you getting?

Just tested and was able to connect to SQL 2014 and 2016 from macOS 13 with OpenSSL 3.1.1.

**Comment by utkalRP** [comment]
> What error message are you getting?

While connecting via my Linux application -> ODBC Driver -> unixODBCManager -> SQL server
https://ibb.co/wcK01Mg

While connecting via openssl -> SQL server
https://ibb.co/8gKj6rz

**Comment by v-chojas** [comment]
This issue is for macOS driver, not Linux one, and it sounds like client and server have no matching ciphersuites so you'll have to figure that out.

**Comment by utkalRP** [comment]
So, MS SQL server 2012/2014/2016 doesn't have OpenSSL 3.0.8 compatible ciphers suits ? OpenSSL have TLSv1, 1.1, 1.2 & 1.3 support. The application built with OpenSSL 1.1.1 is working fine and making a TLSv1.2 connection with same server. But when the application built with OpenSSL 3.0.8, it has the issue!!!

I'm confused where is the issue  :(
> C++ client application   -OR-
> OpenSSL 3.0.8    -OR-
> unixODBC manager    -OR-
> MS SQL ODBC Driver 17/18    -OR-
> Windows host OS : Windows Server 2012 R2    -OR-
> Windows SQL Server 2012/2014/1016
